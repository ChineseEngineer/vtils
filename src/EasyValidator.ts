import { isChineseIDCardNumber, isEmail, isFunction, isInteger, isNumeric, isPossibleChineseMobilePhoneNumber, isPossibleChineseName, isPromiseLike, isRegExp, isUrl } from './is'
import { sequential } from './sequential'

export type EasyValidatorData = Record<keyof any, any>

export interface EasyValidatorRuleTestFunctionActions {
  /**
   * 更新提示信息。
   *
   * @param message 提示信息
   */
  updateMessage(message: any): void,
}

export interface EasyValidatorRuleTestFunction<D extends EasyValidatorData> {
  /**
   * 测试函数，除非显式返回 `false`，否则都将视为测试通过。
   *
   * @param data 数据对象
   * @param actions 操作列表
   * @returns 返回是否测试通过
   */
  (
    data: D,
    actions: EasyValidatorRuleTestFunctionActions,
  ): any,
}

export interface EasyValidatorRule<D extends EasyValidatorData> {
  /** 要验证字段在数据对象中的键名 */
  key: keyof D,
  /** 验证类型 */
  type?: (
    'number' |
    'integer' |
    'chineseMobilePhoneNumber' |
    'chineseIdCardNumber' |
    'url' |
    'email' |
    'chineseName'
  ),
  /** 是否必填 */
  required?: boolean,
  /**
   * 自定义测试，支持正则、同步函数、异步函数。
   *
   * 为函数时，除非显式返回 `false`，否则都将视为测试通过。
   */
  test?: (
    RegExp |
    EasyValidatorRuleTestFunction<D>
  ),
  /** 提示信息 */
  message: any,
}

export type EasyValidatorRules<D extends EasyValidatorData> = Array<EasyValidatorRule<D>>

export interface EasyValidatorValidateReturn<D extends EasyValidatorData> {
  /** 是否验证通过 */
  valid: boolean,
  /** 未验证通过的规则组成的列表 */
  unvalidRules: Array<EasyValidatorRule<D>>,
}

/**
 * 数据对象验证器。
 *
 * @template D 要验证的数据对象类型
 * @example
 * ```ts
 * interface Data {
 *   name: string,
 *   phoneNumber: string,
 *   pass1: string,
 *   pass2: string,
 * }
 * const ev = new EasyValidator<Data>([
 *   {
 *     key: 'name',
 *     type: 'chineseName',
 *     message: '请输入真实姓名',
 *   },
 *   {
 *     key: 'phoneNumber',
 *     type: 'chineseMobilePhoneNumber',
 *     message: '请输入正确的手机号码',
 *   },
 *   {
 *     key: 'phoneNumber',
 *     test: async ({ phoneNumber }, { updateMessage }) => {
 *       const result = await checkPhoneNumberAsync(phoneNumber)
 *       if (!result.valid) {
 *         updateMessage(result.message)
 *         return false
 *       }
 *     },
 *     message: '请输入正确的手机号码'
 *   },
 *   {
 *     key: 'pass1',
 *     test: ({ pass1 }) => pass1.length > 6,
 *     message: '密码应大于6位',
 *   },
 *   {
 *     key: 'pass2',
 *     test: ({ pass1, pass2 }) => pass2 === pass1,
 *     message: '两次密码应一致',
 *   },
 * ])
 * ev.validate({
 *   name: '方一一',
 *   phoneNumber: '18087030070',
 *   pass1: '1234567',
 *   pass2: '12345678'
 * }).then(res => {
 *   // => { valid: false, unvalidRules: [{ key: 'pass2', test: ({ pass1, pass2 }) => pass2 === pass1, message: '两次密码应一致' }] }
 * })
 * ```
 */
export class EasyValidator<D extends EasyValidatorData> {
  /**
   * 构造函数。
   *
   * @param rules 规则列表
   */
  constructor(private rules: EasyValidatorRules<D>) {}

  private check(rule: EasyValidatorRule<D>, data: D) {
    return new Promise<{ valid: boolean, message?: any }>(resolve => {
      const key = rule.key
      const value = data[key]

      /* istanbul ignore if  */
      if (!(key in data)) {
        if (rule.required) {
          return resolve({ valid: false })
        }
      } else {
        if ((rule.required || rule.type || rule.test) && (value == null || value === '')) {
          return resolve({ valid: false })
        }
        if (rule.type) {
          switch (rule.type) {
            case 'number':
              if (!isNumeric(value)) return resolve({ valid: false })
              break
            case 'integer':
              if (!isNumeric(value) || !isInteger(Number(value))) return resolve({ valid: false })
              break
            case 'chineseMobilePhoneNumber':
              if (!isPossibleChineseMobilePhoneNumber(value)) return resolve({ valid: false })
              break
            case 'chineseIdCardNumber':
              if (!isChineseIDCardNumber(value)) return resolve({ valid: false })
              break
            case 'url':
              if (!isUrl(value)) return resolve({ valid: false })
              break
            case 'email':
              if (!isEmail(value)) return resolve({ valid: false })
              break
            case 'chineseName':
              if (!isPossibleChineseName(value)) return resolve({ valid: false })
              break
            /* istanbul ignore next */
            default:
              break
          }
        }
        if (rule.test) {
          if (isRegExp(rule.test)) {
            if (!rule.test.test(value)) {
              return resolve({ valid: false })
            }
          } else if (isFunction(rule.test)) {
            let message: any
            const actions: EasyValidatorRuleTestFunctionActions = {
              updateMessage(comingMessage) {
                message = comingMessage
              },
            }
            const result = rule.test(data, actions)
            if (isPromiseLike(result)) {
              return result
                .then(
                  pass => ({ valid: pass !== false, message }),
                  message => ({ valid: false, message }),
                )
                .then(resolve)
            }
            return resolve({ valid: result !== false, message })
          }
        }
      }

      return resolve({ valid: true })
    })
  }

  /**
   * 验证数据。
   *
   * @param data 要验证的数据
   * @returns 返回验证结果
   */
  validate(data: D) {
    const unvalidKeys: Array<keyof D> = []
    const unvalidRules: Array<EasyValidatorRule<D>> = []

    return (
      sequential(
        this.rules.map(
          rule => () => {
            return new Promise(resolve => {
              if (unvalidKeys.indexOf(rule.key) === -1) {
                this.check(rule, data).then(({ valid, message }) => {
                  if (!valid) {
                    unvalidKeys.push(rule.key)
                    unvalidRules.push(
                      message == null
                        ? rule
                        : { ...rule, message },
                    )
                  }
                  resolve()
                })
              } else {
                resolve()
              }
            })
          },
        ),
      ).then<EasyValidatorValidateReturn<D>>(() => ({
        valid: unvalidRules.length === 0,
        unvalidRules: unvalidRules.slice(),
      }))
    )
  }
}
