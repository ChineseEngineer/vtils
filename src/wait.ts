/**
 * 等待 n 毫秒。
 *
 * @param milliseconds 等待时间，单位：毫秒
 */
export function wait(milliseconds: number): Promise<void> {
  return new Promise(
    resolve => setTimeout(resolve, milliseconds),
  )
}
