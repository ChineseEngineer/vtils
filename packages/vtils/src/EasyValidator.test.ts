import {EasyValidator, EasyValidatorRules} from './EasyValidator'
import {wait} from './wait'

type Data = {
  number?: string,
  integer?: string,
  chineseMobilePhoneNumber?: string,
  chineseIdCardNumber?: string,
  url?: string,
  email?: string,
  chineseName?: string,
  customRegExp?: string,
  customSyncFn?: string,
  customAsyncFn?: string,
  pass1?: string,
  pass2?: string,
  required: string,
  updateMessage: string,
}

test('综合测试', async () => {
  const rules: EasyValidatorRules<Data> = [
    {
      key: 'number',
      type: 'number',
      message: '请输入数字',
    },
    {
      key: 'integer',
      type: 'integer',
      message: '请输入正整数',
    },
    {
      key: 'chineseMobilePhoneNumber',
      type: 'chineseMobilePhoneNumber',
      message: '请输入手机号码',
    },
    {
      key: 'chineseIdCardNumber',
      type: 'chineseIdCardNumber',
      message: '请输入身份证号',
    },
    {
      key: 'url',
      type: 'url',
      message: '请输入网址',
    },
    {
      key: 'email',
      type: 'email',
      message: '请输入邮箱',
    },
    {
      key: 'email',
      test: /@163\.com$/,
      message: '请输入网易邮箱',
    },
    {
      key: 'chineseName',
      type: 'chineseName',
      message: '请输入姓名',
    },
    {
      key: 'customRegExp',
      test: /abc/,
      message: '请输入包含abc的文字',
    },
    {
      key: 'customSyncFn',
      test: ({customSyncFn}) => /abc/.test(customSyncFn!),
      message: '请输入包含abc的文字',
    },
    {
      key: 'customAsyncFn',
      test: async ({customAsyncFn}) => {
        await wait(500)
        return /abc/.test(customAsyncFn!)
      },
      message: '请输入包含abc的文字',
    },
    {
      key: 'pass1',
      required: true,
      test: ({pass1}) => pass1!.length > 6,
      message: '请输入大于6位的密码',
    },
    {
      key: 'pass2',
      test: ({pass1, pass2}) => pass2 === pass1,
      message: '请输入和密码一相同的密码',
    },
    {
      key: 'required',
      required: true,
      message: '请输入',
    },
    {
      key: 'updateMessage',
      test: (_, {updateMessage}) => {
        if (_.updateMessage !== 'x') {
          updateMessage('出错啦')
          return false
        }
      },
      message: 'x',
    },
  ]

  const v = new EasyValidator<Data>(rules)

  expect(
    await v.validate({
      number: '122.3',
      integer: '12',
      chineseMobilePhoneNumber: '18842611520',
      chineseIdCardNumber: '130401200101011678',
      chineseName: '方剑成',
      url: 'http://github.com',
      email: 'fjc@163.com',
      customRegExp: 'abc2',
      customSyncFn: '2abc',
      customAsyncFn: '2abc2',
      pass1: '1234567',
      pass2: '1234567',
      required: 'req',
      updateMessage: 'x',
    }),
  ).toMatchSnapshot('验证成功')

  expect(
    await v.validate({
      number: '122.3',
      integer: '12.2',
      chineseMobilePhoneNumber: '18842611520',
      chineseIdCardNumber: '130401200101011678',
      url: 'http://github.com',
      email: 'fjc@163',
      chineseName: '方剑成',
      customRegExp: 'abd2',
      customSyncFn: '2abc',
      customAsyncFn: '2abs2',
      pass1: '1234567',
      pass2: '12345678',
      required: undefined as any,
      updateMessage: 'y',
    }),
  ).toMatchSnapshot('验证失败')
})

test('示例', async () => {
  async function checkPhoneNumberAsync(_num: string) {
    await wait(500)
    return {
      valid: true,
      message: '',
    }
  }
  interface Data {
    name: string,
    phoneNumber: string,
    pass1: string,
    pass2: string,
  }
  const ev = new EasyValidator<Data>([
    {
      key: 'name',
      type: 'chineseName',
      message: '请输入真实姓名',
    },
    {
      key: 'phoneNumber',
      type: 'chineseMobilePhoneNumber',
      message: '请输入正确的手机号码',
    },
    {
      key: 'phoneNumber',
      test: async ({phoneNumber}, {updateMessage}) => {
        const result = await checkPhoneNumberAsync(phoneNumber)
        if (!result.valid) {
          updateMessage(result.message)
          return false
        }
      },
      message: '请输入正确的手机号码',
    },
    {
      key: 'pass1',
      test: ({pass1}) => pass1.length > 6,
      message: '密码应大于6位',
    },
    {
      key: 'pass2',
      test: ({pass1, pass2}) => pass2 === pass1,
      message: '两次密码应一致',
    },
  ])
  const res = await ev.validate({
    name: '方一一',
    phoneNumber: '18087030070',
    pass1: '1234567',
    pass2: '12345678',
  })
  expect(res).toMatchSnapshot('示例')
})
