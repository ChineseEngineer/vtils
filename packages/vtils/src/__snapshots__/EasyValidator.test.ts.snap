// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`示例: 示例 1`] = `
Object {
  "firstUnvalidRuleMessage": "两次密码应一致",
  "messages": Object {
    "pass2": "两次密码应一致",
  },
  "unvalidRules": Array [
    Object {
      "key": "pass2",
      "message": "两次密码应一致",
      "test": [Function],
    },
  ],
  "valid": false,
  "validByKey": Object {
    "name": true,
    "pass1": true,
    "pass2": false,
    "phoneNumber": true,
  },
}
`;

exports[`综合测试: 验证失败 1`] = `
Object {
  "firstUnvalidRuleMessage": "请输入正整数",
  "messages": Object {
    "customAsyncFn": "请输入包含abc的文字",
    "customRegExp": "请输入包含abc的文字",
    "email": "请输入邮箱",
    "integer": "请输入正整数",
    "pass2": "请输入和密码一相同的密码",
    "required": "请输入",
    "updateMessage": "出错啦",
  },
  "unvalidRules": Array [
    Object {
      "key": "integer",
      "message": "请输入正整数",
      "type": "integer",
    },
    Object {
      "key": "email",
      "message": "请输入邮箱",
      "type": "email",
    },
    Object {
      "key": "customRegExp",
      "message": "请输入包含abc的文字",
      "test": /abc/,
    },
    Object {
      "key": "customAsyncFn",
      "message": "请输入包含abc的文字",
      "test": [Function],
    },
    Object {
      "key": "pass2",
      "message": "请输入和密码一相同的密码",
      "test": [Function],
    },
    Object {
      "key": "required",
      "message": "请输入",
      "required": true,
    },
    Object {
      "key": "updateMessage",
      "message": "出错啦",
      "test": [Function],
    },
  ],
  "valid": false,
  "validByKey": Object {
    "chineseIdCardNumber": true,
    "chineseMobilePhoneNumber": true,
    "chineseName": true,
    "customAsyncFn": false,
    "customRegExp": false,
    "customSyncFn": true,
    "email": false,
    "integer": false,
    "number": true,
    "pass1": true,
    "pass2": false,
    "required": false,
    "updateMessage": false,
    "url": true,
  },
}
`;

exports[`综合测试: 验证成功 1`] = `
Object {
  "firstUnvalidRuleMessage": undefined,
  "messages": Object {},
  "unvalidRules": Array [],
  "valid": true,
  "validByKey": Object {
    "chineseIdCardNumber": true,
    "chineseMobilePhoneNumber": true,
    "chineseName": true,
    "customAsyncFn": true,
    "customRegExp": true,
    "customSyncFn": true,
    "email": true,
    "integer": true,
    "number": true,
    "pass1": true,
    "pass2": true,
    "required": true,
    "updateMessage": true,
    "url": true,
  },
}
`;
